File: MAStaff-API/src/main/java/es/angelillo15/mast/api/cmd/CommandParent.java
package es.angelillo15.mast.api.cmd;

import es.angelillo15.mast.api.TextUtils;
import es.angelillo15.mast.api.cmd.sender.CommandSender;
import lombok.Getter;
import lombok.Setter;

import java.util.HashMap;
import java.util.Map;

public abstract class CommandParent extends Command {
    @Getter
    private final Map<String, SubCommand> subCommands = new HashMap<>();
    @Getter
    @Setter
    private static String noPermission = "&cYou don't have permission to execute this command.";

    @Override
    public void onCommand(CommandSender sender, String label, String[] args) {
        if (args.length == 0) {
            sendHelp(sender);
            return;
        }

        SubCommand subCommand = getSubCommand(args[0]);
        if (subCommand == null) {
            sendHelp(sender);
            return;
        }

        if (!sender.hasPermission(subCommand.getPermission())) {
            sender.sendMessage(TextUtils.simpleColorize(noPermission));
            return;
        }

        subCommand.onCommand(sender, label, args);
    }

    public void sendHelp(CommandSender sender) {
        sender.sendMessage(TextUtils.simpleColorize("&6----------------MAStaff----------------"));
        sender.sendMessage(TextUtils.simpleColorize("&bAvailable Commands:"));
        for (SubCommand subCommand : subCommands.values()) {
            sender.sendMessage(TextUtils.simpleColorize("&b" + subCommand.getSyntax() + " &7- &f" + subCommand.getDescription()));
        }
    }

    public void registerSubCommand(SubCommand subCommand) {
        subCommands.put(subCommand.getName(), subCommand);
    }

    public void unregisterSubCommand(SubCommand subCommand) {
        subCommands.remove(subCommand.getName());
    }

    public void unregisterSubCommand(String name) {
        subCommands.remove(name);
    }

    public SubCommand getSubCommand(String name) {
        return subCommands.get(name);
    }

    public boolean hasSubCommand(String name) {
        return subCommands.containsKey(name);
    }
}

File: MAStaff-API/src/main/java/es/angelillo15/mast/api/cmd/CommandParent.kt
package es.angelillo15.mast.api.cmd

import es.angelillo15.mast.api.Constants
import es.angelillo15.mast.api.TextUtils
import es.angelillo15.mast.api.cmd.sender.CommandSender
import lombok.Getter
import lombok.Setter

abstract class CommandParent : Command() {
    @Getter
    private val subCommands: MutableMap<String?, SubCommand> = HashMap()
    override fun onCommand(sender: CommandSender?, label: String?, args: Array<String?>?) {
        if (subCommands.isEmpty())
            registerSubCommands()

        if (args!!.isEmpty()) {
            sendHelp(sender)
            return
        }
        val subCommand = getSubCommand(args[0])
        if (subCommand == null) {
            sendHelp(sender)
            return
        }
        if (!sender!!.hasPermission(subCommand.permission)) {
            sender.sendMessage(TextUtils.simpleColorize(noPermission))
            return
        }

        subCommand.onCommand(sender, label, args)
    }

    abstract fun registerSubCommands();

    fun sendHelp(sender: CommandSender?) {
        sender!!.sendMessage(TextUtils.simpleColorize("&a&lMAS&r&ltaff &7- &fv" + Constants.VERSION))
        for (subCommand in subCommands.values) {
            if (sender.hasPermission(subCommand.permission)) {
                sender.sendMessage(TextUtils.simpleColorize("&a&l> &r" + subCommand.syntax + " &7- &7" + subCommand.description))
            }
        }
    }

    fun registerSubCommand(subCommand: SubCommand) {
        subCommands[subCommand.name] = subCommand
    }

    fun unregisterSubCommand(subCommand: SubCommand) {
        subCommands.remove(subCommand.name)
    }

    fun unregisterSubCommand(name: String?) {
        subCommands.remove(name)
    }

    fun getSubCommand(name: String?): SubCommand? {
        return subCommands[name]
    }

    fun hasSubCommand(name: String?): Boolean {
        return subCommands.containsKey(name)
    }

    companion object {
        @Getter
        @Setter
        private val noPermission = "&cYou don't have permission to execute this command."
    }
}
